sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala

```diff
@@ -1917,19 +1921,33 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSparkSession {
       null
     ).toDF("i")
 
+    // transform(i, x -> x + 1)
+    val resA = Seq(
+      Row(Seq(2, 10, 9, 8)),
+      Row(Seq(6, 9, 10, 8, 3)),
+      Row(Seq.empty),
+      Row(null))
+
+    // transform(i, (x, i) -> x + i)
+    val resB = Seq(
+      Row(Seq(1, 10, 10, 10)),
+      Row(Seq(5, 9, 11, 10, 6)),
+      Row(Seq.empty),
+      Row(null))
+
     def testArrayOfPrimitiveTypeNotContainsNull(): Unit = {
-      checkAnswer(df.selectExpr("transform(i, x -> x + 1)"),
-        Seq(
-          Row(Seq(2, 10, 9, 8)),
-          Row(Seq(6, 9, 10, 8, 3)),
-          Row(Seq.empty),
-          Row(null)))
-      checkAnswer(df.selectExpr("transform(i, (x, i) -> x + i)"),
-        Seq(
-          Row(Seq(1, 10, 10, 10)),
-          Row(Seq(5, 9, 11, 10, 6)),
-          Row(Seq.empty),
-          Row(null)))
+      checkAnswer(df.selectExpr("transform(i, x -> x + 1)"), resA)
+      checkAnswer(df.selectExpr("transform(i, (x, i) -> x + i)"), resB)
+
+      checkAnswer(df.select(transform(col("i"), x => x + 1)), resA)
+      checkAnswer(df.select(transform(col("i"), (x, i) => x + i)), resB)
+
+      checkAnswer(df.select(transform(col("i"), new JFunc {
+        def call(x: Column) = x + 1
+      })), resA)
+      checkAnswer(df.select(transform(col("i"), new JFunc2 {
+        def call(x: Column, i: Column) = x + i
+      })), resB)
```

Comment:
Could you move Java tests to `JavaDataFrameSuite` or create `JavaDataFrameFunctionsSuite`?

Seems like Java compiler claims `error: reference to transform is ambiguous` because there are two overloaded functions `def transform(column: Column, f: Column => Column): Column` and `def transform(column: Column, f: JavaFunction[Column, Column])`, and the same for the others.

I guess Java compiler can handle `Column => Column` so we don't need the overloaded function with `JavaFunction[Column, Column]`.

Reply:
Sure, I'll try that out.

Reply:
That's strange though since Java compiler can handle the overloaded `to_json` functions, one that takes a scala map and one that takes a java map

Reply:
Can you share the code that gave the error @ueshin ?

Reply:
```java
transform(col("arr"), x -> x);
```

Reply:
The situations between here and `to_json` are not the same.

As for `to_json`, their arguments are `scala.collection.Map` and `java.util.Map`, so both are needed for the convenience for both Scala and Java users.

As for these, the arguments are `scala.Function1` and `org.apache.spark.api.java.function.Function`, Java compiler recognizes both as SAM (Single Abstract Method) type when resolving `x -> x`, and they are ambiguous.

Reply:
I'm able to compile this snippet without error.
```java
import org.apache.spark.api.java.function.*;
import org.apache.spark.sql.Column;
import static org.apache.spark.sql.functions.*;

public class Foo {
  public Column x = transform(expr("array(1, 2, 3)"), new Function<Column, Column>() {
    public Column call(Column x) {
        return x;
    }
  });
}

```

Reply:
Yeah, it's because you explicitly specify `new Function<Column, Column>() ...`.
Users usually use lambda expressions like `x -> x`, then it will fail.

Reply:
Please try this:

```java

import org.junit.*;

import static org.apache.spark.sql.functions.*;

public class JavaDataFrameFunctionsSuite {
  @Test
  public void testExecution() {
    transform(expr("array(1, 2, 3)"), x -> x);
  }
}
```

Reply:
Finished adding tests for java overload to the DataFrameFunctionsSuite. If it turns out we need to move it after investigating your error @ueshin, then I can do that too.

Reply:
The Java API tests you added to `DataFrameFunctionsSuite` are not useful.
Please move them to `JavaDataFrameSuite` or create `JavaDataFrameFunctionsSuite` with lambda expressions as users usually do like above and see what happens.

Reply:
@nvander1, let's move Java specific tests to Java's. Ideally it should be moved to there and it's better to clarify such doubts.

Reply:
> As for these, the arguments are scala.Function1 and org.apache.spark.api.java.function.Function, Java compiler recognizes both as SAM (Single Abstract Method) type when resolving x -> x, and they are ambiguous.

Sounds like a valid concern to me. Does such `x -> x` syntax work in Java too?

Reply:
I did a quick experiement and @ueshin is right, the `x -> x` syntax correctly passes to functions expecting `x => x`

Interesting that the java compiler allows passing the Java8 lambdas to functions expecting scala lambdas. I had assumed since the scala `Function` are implemented as traits, and not the java `@FunctionalInterface` it wouldn't work. Do you know why it works @ueshin?

I'll work on removing the overloads with `JavaFunction` and make tests that use the `x -> x` syntax in Java. :)

Reply:
IIUC:
1. The Java compiler tries to resolve lambdas as a Single Abstract Method (SAM) type.
2. The SAM type has only one abstract method and can have other non-abstract methods including default methods for interfaces. Not only interfaces but also abstract classes can be a SAM type.
3. `@FunctionalInterface` is just a marker for Java compiler which implies the annotated interface or class is the SAM type. SAM types are not necessarily annotated with `@FunctionalInterface`.
4. The scala `Function` satisfies the SAM type requirements:
   -  Scala's trait is basically an interface.
   - `Function` has only one `apply()` method.

Reply:
Thanks for the explanation!